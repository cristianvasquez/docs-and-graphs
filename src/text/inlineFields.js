const PARENTHESIS_TYPES = [
  {
    left: '(', right: ')',
  }, {
    left: '[', right: ']',
  }]

function keyValue ({ content, parenthesis }) {
  const chunks = content.split('::')
  if (chunks.length === 2) {
    return {
      property: chunks[0],
      value: chunks[1],
      raw: `${parenthesis.left}${content}${parenthesis.right}`,
    }
  }
  return undefined
}

function parseParenthesis (str) {
  const result = []
  for (const parenthesis of PARENTHESIS_TYPES) {
    for (const [i, value] of str.split(parenthesis.left).entries()) {
      if (i > 0) {
        const content = value.split(parenthesis.right)[0]
        result.push({
          content, parenthesis,
        })
      }
    }
  }
  return result
}

function parseWithoutParenthesis (str) {
  return [
    keyValue({
      content: str, parenthesis: {
        left: '', right: '',
      },
    })].filter(valid => valid)
}

function parseWithParenthesis (str) {
  return parseParenthesis(str).map(keyValue).filter(valid => valid)
}

// Accepts embedded pairs between brackets, [like:: this] (or like::this)
function extractInlineFields (str) {
  const content = parseWithParenthesis(str)
  return content.length ? content : parseWithoutParenthesis(str)
}

function removeInlineFields (str) {
  let result = str
  for (const current of extractInlineFields(str)) {
    result = result.replaceAll(current.raw, '')
  }
  return result
}

export { extractInlineFields, removeInlineFields }
