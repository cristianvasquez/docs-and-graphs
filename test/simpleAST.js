import { expect } from 'expect'
import { simpleAst } from '../index.js'
import toMatchSnapshot from 'expect-mocha-snapshot'
import tests from './tests.js'
import inlineFields from './support/inlineFields.js'
import maxDepth from './support/maxDepth.js'

expect.extend({ toMatchSnapshot })

describe('astDag', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const fullText = current.markdown

      const result = simpleAst(fullText)

      expect(result).toMatchSnapshot(this)
    })
  }
})

describe('astDag normalize:true', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const fullText = current.markdown

      const result = simpleAst(fullText, { normalize: true })

      expect(result).toMatchSnapshot(this)
    })
  }
})

describe('astDag includePosition:true', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const fullText = current.markdown
      const result = simpleAst(fullText, { includePosition: true })
      expect(result).toMatchSnapshot(this)
    })
  }
})

describe('inLineAsArray', async function () {
  it('inLineAsArray:true', async function () {
    const fullText = inlineFields.markdown
    const result = simpleAst(fullText, { inlineAsArray: true })

    expect(result).toMatchSnapshot(this)
  })
  it('inLineAsArray:false', async function () {
    const fullText = inlineFields.markdown
    const result = simpleAst(fullText, { inlineAsArray: false })

    expect(result).toMatchSnapshot(this)
  })
})

describe('maxDepth', async function () {
  it('maxDepth:2', async function () {
    const fullText = maxDepth.markdown
    const result = simpleAst(fullText, { maxDepth: 2 })

    expect(result).toMatchSnapshot(this)
  })
  it('maxDepth:3', async function () {
    const fullText = maxDepth.markdown
    const result = simpleAst(fullText, { maxDepth: 3 })

    expect(result).toMatchSnapshot(this)
  })
  it('no maxDepth (default)', async function () {
    const fullText = maxDepth.markdown
    const result = simpleAst(fullText)

    expect(result).toMatchSnapshot(this)
  })
})
