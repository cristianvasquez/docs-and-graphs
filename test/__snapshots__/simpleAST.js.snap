// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`astDag Headings 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Some text",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "H2",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "H1",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag Inline fields 1`] = `
Object {
  "children": Array [
    Object {
      "data": Array [
        Object {
          "inline ": "field",
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          " inline ": "field",
        },
      ],
      "type": "text",
      "value": "nested",
    },
    Object {
      "data": Array [
        Object {
          "**Thoughts**": "It was decent.",
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          "**Rating**": "6'",
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          "mood": "okay",
        },
        Object {
          "length": "2 hours",
        },
      ],
      "type": "text",
      "value": "|",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " inline ": "field",
          },
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          " subject ": Object {
            " inline ": "field",
          },
        },
      ],
      "type": "text",
      "value": "nested",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " **Thoughts**": "It was decent.",
          },
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " **Rating**": "6",
          },
        },
      ],
      "type": "text",
      "value": "",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " mood": "okay",
          },
        },
        Object {
          "subject ": Object {
            " length": "2 hours",
          },
        },
      ],
      "type": "text",
      "value": "|",
    },
    Object {
      "type": "text",
      "value": "No inline fields",
    },
    Object {
      "data": Array [
        Object {
          "Too ": Object {
            " many ": Object {
              " inline ": "fields",
            },
          },
        },
      ],
      "type": "text",
      "value": "",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag Links 1`] = `
Object {
  "children": Array [
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "link 1",
        },
      ],
      "type": "text",
      "value": "[[link 1]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "/full/path/link 1",
        },
      ],
      "type": "text",
      "value": "[[/full/path/link 1]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "/full/path/link 1.md",
        },
      ],
      "type": "text",
      "value": "[[/full/path/link 1.md]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": "Alias 2",
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "[Alias 2](http://example.com)",
    },
    Object {
      "links": Array [
        Object {
          "alias": "Alias 1",
          "type": "wikiLink",
          "value": "link 1 ",
        },
      ],
      "type": "text",
      "value": "[[link 1 |Alias 1]]",
    },
    Object {
      "type": "text",
      "value": "![[link 1]]",
    },
    Object {
      "type": "text",
      "value": "![[image.png]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "The website is http://example.com see you!",
    },
    Object {
      "links": Array [
        Object {
          "alias": "here",
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "The website is [here](http://example.com) see you!",
    },
    Object {
      "links": Array [
        Object {
          "type": "link",
          "value": "protocol",
        },
      ],
      "type": "text",
      "value": "The website is [here](protocol) see you!",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag Outline 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "outline",
          "value": "Task",
        },
        Object {
          "checked": true,
          "type": "outline",
          "value": "Task",
        },
        Object {
          "type": "outline",
          "value": "Item 0",
        },
        Object {
          "children": Array [
            Object {
              "type": "outline",
              "value": "Item 1",
            },
            Object {
              "type": "outline",
              "value": "Item 2",
            },
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Item 3",
                },
              ],
              "ordered": false,
              "type": "outline",
            },
          ],
          "ordered": false,
          "type": "outline",
        },
        Object {
          "type": "outline",
          "value": "Item 4",
        },
      ],
      "ordered": false,
      "type": "outline",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag People 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "data": Array [
                Object {
                  "number ": "42",
                },
              ],
              "type": "text",
              "value": "",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "A heading outside",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "Alice",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Eating",
                },
                Object {
                  "type": "outline",
                  "value": "Dancing",
                },
              ],
              "ordered": false,
              "type": "outline",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "Interests",
        },
      ],
      "data": Array [
        Object {
          "age ": "42",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "Bob",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Waking up",
                },
                Object {
                  "type": "outline",
                  "value": "Drinking tea",
                },
              ],
              "ordered": true,
              "type": "outline",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "Todo list",
        },
        Object {
          "children": Array [
            Object {
              "data": Array [
                Object {
                  "age ": "56",
                },
              ],
              "type": "text",
              "value": "Charlie is a good guy.",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "Description",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "Charlie",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag Yaml 1`] = `
Object {
  "data": Array [
    Object {
      "date updated": "2021-08-05T17:10:34+02:00",
      "one": Array [
        "Two",
        Object {
          "Three": Array [
            "Other",
          ],
        },
      ],
    },
  ],
  "depth": 0,
  "tags": Array [
    "cats",
    "dogs",
  ],
  "type": "root",
}
`;

exports[`astDag normalize:false Headings 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Some text",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "## H2",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "# H1",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag normalize:false Inline fields 1`] = `
Object {
  "children": Array [
    Object {
      "data": Array [
        Object {
          "inline ": " field",
        },
      ],
      "type": "text",
      "value": "inline :: field",
    },
    Object {
      "data": Array [
        Object {
          " inline ": " field ",
        },
      ],
      "type": "text",
      "value": "nested ( inline :: field )",
    },
    Object {
      "data": Array [
        Object {
          "**Thoughts**": " It was decent.",
        },
      ],
      "type": "text",
      "value": "**Thoughts**:: It was decent.",
    },
    Object {
      "data": Array [
        Object {
          "**Rating**": " 6'",
        },
      ],
      "type": "text",
      "value": "**Rating**:: 6'",
    },
    Object {
      "data": Array [
        Object {
          "mood": " okay",
        },
        Object {
          "length": " 2 hours",
        },
      ],
      "type": "text",
      "value": "[mood:: okay] | [length:: 2 hours]",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " inline ": " field",
          },
        },
      ],
      "type": "text",
      "value": "subject :: inline :: field",
    },
    Object {
      "data": Array [
        Object {
          " subject ": Object {
            " inline ": " field ",
          },
        },
      ],
      "type": "text",
      "value": "nested ( subject :: inline :: field )",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " **Thoughts**": " It was decent.",
          },
        },
      ],
      "type": "text",
      "value": "subject :: **Thoughts**:: It was decent.",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " **Rating**": " 6",
          },
        },
      ],
      "type": "text",
      "value": "subject :: **Rating**:: 6",
    },
    Object {
      "data": Array [
        Object {
          "subject ": Object {
            " mood": " okay",
          },
        },
        Object {
          "subject ": Object {
            " length": " 2 hours",
          },
        },
      ],
      "type": "text",
      "value": "[subject :: mood:: okay] | [subject :: length:: 2 hours]",
    },
    Object {
      "type": "text",
      "value": "No inline fields",
    },
    Object {
      "data": Array [
        Object {
          "Too ": Object {
            " many ": Object {
              " inline ": " fields",
            },
          },
        },
      ],
      "type": "text",
      "value": "Too :: many :: inline :: fields",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag normalize:false Links 1`] = `
Object {
  "children": Array [
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "link 1",
        },
      ],
      "type": "text",
      "value": "[[link 1]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "/full/path/link 1",
        },
      ],
      "type": "text",
      "value": "[[/full/path/link 1]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "wikiLink",
          "value": "/full/path/link 1.md",
        },
      ],
      "type": "text",
      "value": "[[/full/path/link 1.md]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": "Alias 2",
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "[Alias 2](http://example.com)",
    },
    Object {
      "links": Array [
        Object {
          "alias": "Alias 1",
          "type": "wikiLink",
          "value": "link 1 ",
        },
      ],
      "type": "text",
      "value": "[[link 1 |Alias 1]]",
    },
    Object {
      "type": "text",
      "value": "![[link 1]]",
    },
    Object {
      "type": "text",
      "value": "![[image.png]]",
    },
    Object {
      "links": Array [
        Object {
          "alias": undefined,
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "The website is http://example.com see you!",
    },
    Object {
      "links": Array [
        Object {
          "alias": "here",
          "type": "link",
          "value": "http://example.com",
        },
      ],
      "type": "text",
      "value": "The website is [here](http://example.com) see you!",
    },
    Object {
      "links": Array [
        Object {
          "type": "link",
          "value": "protocol",
        },
      ],
      "type": "text",
      "value": "The website is [here](protocol) see you!",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag normalize:false Outline 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "outline",
          "value": "Task",
        },
        Object {
          "checked": true,
          "type": "outline",
          "value": "Task",
        },
        Object {
          "type": "outline",
          "value": "Item 0",
        },
        Object {
          "children": Array [
            Object {
              "type": "outline",
              "value": "Item 1",
            },
            Object {
              "type": "outline",
              "value": "Item 2",
            },
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Item 3",
                },
              ],
              "ordered": false,
              "type": "outline",
            },
          ],
          "ordered": false,
          "type": "outline",
        },
        Object {
          "type": "outline",
          "value": "Item 4",
        },
      ],
      "ordered": false,
      "type": "outline",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag normalize:false People 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "data": Array [
                Object {
                  "number ": " 42",
                },
              ],
              "type": "text",
              "value": "number :: 42",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "## A heading outside",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "# Alice",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Eating",
                },
                Object {
                  "type": "outline",
                  "value": "Dancing",
                },
              ],
              "ordered": false,
              "type": "outline",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "## Interests",
        },
      ],
      "data": Array [
        Object {
          "age ": " 42",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "# Bob (age :: 42)",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "outline",
                  "value": "Waking up",
                },
                Object {
                  "type": "outline",
                  "value": "Drinking tea",
                },
              ],
              "ordered": true,
              "type": "outline",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "## Todo list",
        },
        Object {
          "children": Array [
            Object {
              "data": Array [
                Object {
                  "age ": " 56",
                },
              ],
              "type": "text",
              "value": "Charlie is a good guy. (age :: 56)",
            },
          ],
          "depth": 2,
          "type": "block",
          "value": "## Description",
        },
      ],
      "depth": 1,
      "type": "block",
      "value": "# Charlie",
    },
  ],
  "depth": 0,
  "type": "root",
}
`;

exports[`astDag normalize:false Yaml 1`] = `
Object {
  "data": Array [
    Object {
      "date updated": "2021-08-05T17:10:34+02:00",
      "one": Array [
        "Two",
        Object {
          "Three": Array [
            "Other",
          ],
        },
      ],
    },
  ],
  "depth": 0,
  "tags": Array [
    "cats",
    "dogs",
  ],
  "type": "root",
}
`;
